lib/stages/BadInputFlag.cpp

```diff
@@ -0,0 +1,135 @@
+#include "BadInputFlag.hpp"
+
+#include "Config.hpp"            // for Config
+#include "Hash.hpp"              // for operator<
+#include "StageFactory.hpp"      // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "buffer.h"              // for allocate_new_metadata_object, mark_frame_empty, mark_fram...
+#include "bufferContainer.hpp"   // for bufferContainer
+#include "datasetManager.hpp"    // for dset_id_t, state_id_t, datasetManager
+#include "datasetState.hpp"      // for prodState
+#include "kotekanLogging.hpp"    // for FATAL_ERROR, WARN
+#include "prometheusMetrics.hpp" // for Gauge, Metrics, MetricFamily
+#include "visBuffer.hpp"         // for VisFrameView, VisField, VisField::vis, VisField::weight
+#include "visUtil.hpp"           // for prod_ctype, frameID, cmap, icmap, modulo, cfloat
+
+#include "gsl-lite.hpp" // for span
+
+#include <algorithm> // for max, binary_search, copy, sort
+#include <atomic>    // for atomic_bool
+#include <cmath>
+#include <complex>      // for complex
+#include <cxxabi.h>     // for __forced_unwind
+#include <exception>    // for exception
+#include <functional>   // for _Bind_helper<>::type, bind, function
+#include <future>       // for future, async
+#include <iterator>     // for back_insert_iterator, back_inserter
+#include <regex>        // for match_results<>::_Base_type
+#include <stdexcept>    // for out_of_range, runtime_error
+#include <stdint.h>     // for uint16_t, uint32_t
+#include <system_error> // for system_error
+#include <utility>      // for pair, tuple_element<>::type
+
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+using kotekan::prometheus::Metrics;
+
+REGISTER_KOTEKAN_STAGE(BadInputFlag);
+
+BadInputFlag::BadInputFlag(Config& config, const std::string& unique_name,
+                           bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&BadInputFlag::main_thread, this)),
+    bad_input_counter(Metrics::instance().add_counter("kotekan_badinputflag_frames_total",
+                                                      unique_name, {"input_id", "type"})) {
+
+    // Get buffers
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+    out_buf = get_buffer("out_buf");
+    register_producer(out_buf, unique_name.c_str());
+}
+
+void BadInputFlag::check_dataset_state(dset_id_t ds_id) {
+
+    auto& dm = datasetManager::instance();
+
+    auto fprint = dm.fingerprint(ds_id, {"inputs", "products", "stack"});
+
+    if (fingerprint == fingerprint_t::null) {
```

Comment:
Can you explain why we do this? i.e. why do we want to error if there's a stack state or if the inputs/products change?

Reply:
Because it doesn't make sense if the interpretation of the elements in the array changes (which it would if any of those states changed). This allows me to validate once that the states I'm getting makes sense (e.g. the data isn't stacked), and then I never need to do the full validation again, only check the fingerprint.

Reply:
In the code I meant

Reply:
Sorry. I'm not following. Are you just suggesting that I add a comment explaining why this is being done?

Reply:
Yes. To it's clear how the fingerprint works, but not why you choose to error if these things change. Maybe a short explanation in a comment would help someone in the future